src/gateway/adhoc.cpp

```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
 }
 
 auto
-adhoc_t::resolve(const partition_t& name) const -> std::vector<asio::ip::tcp::endpoint> {
-    remote_map_t::const_iterator lb, ub;
-
-    auto ptr = m_remotes.synchronize();
-
-    if(!ptr->count(name)) {
-        throw std::system_error(error::service_not_available);
-    }
-
-    std::tie(lb, ub) = ptr->equal_range(name);
-
-    std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
-    std::advance(lb, distribution(m_random_generator));
-
-    COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", attribute_list({
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> std::shared_ptr<service_description_t> {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        remote_map_t::const_iterator lb, ub;
+        std::tie(lb, ub) = remotes.equal_range(name);
+        if(lb == ub) {
```

Comment:
Really, it took about minute to figure out what's going on here. Return previous condition.